engine/src/main/java/org/pentaho/di/job/entries/shell/JobEntryShell.java

```diff
@@ -0,0 +1,779 @@
+
+      // any output?
+      StreamLogger outputLogger = new StreamLogger( log, proc.getInputStream(), "(stdout)" );
+
+      // kick them off
+      Thread errorLoggerThread = new Thread( errorLogger );
+      errorLoggerThread.start();
+      Thread outputLoggerThread = new Thread( outputLogger );
+      outputLoggerThread.start();
+
+      proc.waitFor();
+      if ( log.isDetailed() ) {
+        logDetailed( BaseMessages.getString( PKG, "JobShell.CommandFinished", command.toString() ) );
+      }
+
+      // What's the exit status?
+      result.setExitStatus( proc.exitValue() );
+      if ( result.getExitStatus() != 0 ) {
+        if ( log.isDetailed() ) {
+          logDetailed( BaseMessages.getString(
+            PKG, "JobShell.ExitStatus", environmentSubstitute( getFilename() ), "" + result.getExitStatus() ) );
+        }
+
+        result.setNrErrors( 1 );
+      }
+
+      // wait until loggers read all data from stdout and stderr
+      errorLoggerThread.join();
+      outputLoggerThread.join();
+
+      // close the streams
+      // otherwise you get "Too many open files, java.io.IOException" after a lot of iterations
+      proc.getErrorStream().close();
+      proc.getOutputStream().close();
+
+    } catch ( IOException ioe ) {
+      logError( BaseMessages.getString(
+        PKG, "JobShell.ErrorRunningShell", environmentSubstitute( getFilename() ), ioe.toString() ), ioe );
+      result.setNrErrors( 1 );
+    } catch ( InterruptedException ie ) {
+      logError( BaseMessages.getString(
+        PKG, "JobShell.Shellinterupted", environmentSubstitute( getFilename() ), ie.toString() ), ie );
+      result.setNrErrors( 1 );
+    } catch ( Exception e ) {
+      logError( BaseMessages.getString( PKG, "JobShell.UnexpectedError", environmentSubstitute( getFilename() ), e
+        .toString() ), e );
+      result.setNrErrors( 1 );
+    } finally {
+      // If we created a temporary file, remove it...
+      //
+      if ( tempFile != null ) {
+        try {
+          tempFile.delete();
+        } catch ( Exception e ) {
+          BaseMessages.getString( PKG, "JobShell.UnexpectedError", tempFile.toString(), e.toString() );
+        }
+      }
+    }
+
+    if ( result.getNrErrors() > 0 ) {
+      result.setResult( false );
+    } else {
+      result.setResult( true );
+    }
+  }
+
+  private FileObject createTemporaryShellFile( FileObject tempFile, String fileContent ) throws Exception {
+    // Create a unique new temporary filename in the working directory, put the script in there
+    // Set the permissions to execute and then run it...
+    //
+    if ( tempFile != null && fileContent != null ) {
+      try {
+        // flag indicates if current OS is Windows or not
+        boolean isWindows = Const.isWindows();
+        if ( !isWindows ) {
+          fileContent = replaceWinEOL( fileContent );
+        }
+        tempFile.createFile();
+        OutputStream outputStream = tempFile.getContent().getOutputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *outputStream*. The resource is referenced at line 675. The resource is closed at line 676. There are other execution paths that do not contain closure statements, for example, when OutputStream.write throws an exception. Ensure *outputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).